### PR TITLE
feat: Add FP4 (E2M1) KV Cache Support for MHA

### PR DESCRIPTION
**Pull request by JackChuang**
# Summary
This PR introduces support for **FP4 (float4_e2m1fn_x2) KV caching** in Multi-Headed Attention (MHA)  e.g., Qwen and GPT-OSS . See #10083, points 1-2, for more context.

Co-authored-by: @yicwang Yichen Wang [yichen.wang@bytedance.com](mailto:yichen.wang@bytedance.com)

# Usage
```
$ python3 -m sglang.launch_server --kv-cache-dtype fp4_e2m1 ... 
```

### **Motivation** and Benefits
Large models often face GPU memory constraints when storing KV cache.
By introducing **FP4 quantization with scale buffers**, this PR significantly reduces KV memory usage and improves efficiency:
- Supports significantly more tokens than KV8 (≈1.78×) and KV16 (≈3.56×) due to FP4 quantization with block_size = 16.
- Improves scalability for longer context windows and throughput for large batch requests
- Enables inference of larger models or longer context windows on memory-limited GPUs.
- Seamless integration with existing inference pipelines without breaking KV16/KV8 workflows.

# Key Changes
- **MHATokenToKVPool**
    - Added FP4 KV cache support with uint8 storage format.
    - Introduced k_scale_buffer and v_scale_buffer for per-block scaling factors.
    - Integrated batched quantization (on update) and dequantization (on access) using KVFP4QuantizeUtil.
- **ModelRunner**
    - Updated GPU memory estimation logic to account for FP4 cache and scale buffers.
- **Compatibility**
    - Preserves existing FP16/FP8 KV cache behavior without changes.

# Accuracy tests for KV4 MHA
- FP4 KV cache is well-suited for large-scale models, providing memory savings with minimal accuracy impact.
- For smaller models, careful evaluation is needed to balance memory efficiency and accuracy.
## Qwen3-235B-A22B

| Model          | Dataset      | Metric   | Subset  | Num  | Score  | Cat.0   |
| -------------- | ------------ | -------- | ------- | ---- | ------ | ------- |
| KV4 (fp4_e2m1) |              |          |         |      |        |         |
| KV4            | gsm8k        | mean_acc | main    | 6595 | 0.9186 | default |
| KV4            | aime25       | mean_acc | OVERALL | 150  | 0.6    | -       |
| KV4            | gpqa_diamond | mean_acc | default | 990  | 0.6778 | default |
| KV8 (fp8_e4m3) |              |          |         |      |        |         |
| KV8            | gsm8k        | mean_acc | main    | 6595 | 0.9181 | default |
| KV8            | aime25       | mean_acc | OVERALL | 150  | 0.7333 | -       |
| KV8            | gpqa_diamond | mean_acc | default | 990  | 0.6899 | default |
| KV16           |              |          |         |      |        |         |
| KV16           | gsm8k        | mean_acc | main    | 6595 | 0.9168 | default |
| KV16           | aime25       | mean_acc | OVERALL | 150  | 0.7733 | -       |
| KV16           | gpqa_diamond | mean_acc | default | 990  | 0.701  | default |

## gpt-oss-120b

| Model          | Dataset      | Metric   | Subset  | Num  | Score  | Cat.0   |
| -------------- | ------------ | -------- | ------- | ---- | ------ | ------- |
| KV4 (fp4_e2m1) |              |          |         |      |        |         |
| KV4            | aime25       | mean_acc | OVERALL | 150  | 0.3533 | -       |
| KV4            | gsm8k        | mean_acc | main    | 6595 | 0.9152 | default |
| KV4            | gpqa_diamond | mean_acc | default | 990  | 0.3202 | default |
| KV8 (fp8_e4m3) |              |          |         |      |        |         |
| KV8            | aime25       | mean_acc | OVERALL | 150  | 0.7667 | -       |
| KV8            | gsm8k        | mean_acc | main    | 6595 | 0.9163 | default |
| KV8            | gpqa_diamond | mean_acc | default | 990  | 0.5434 | default |
| KV16           |              |          |         |      |        |         |
| KV16           | aime25       | mean_acc | OVERALL | 150  | 0.7533 | -       |
| KV16           | gsm8k        | mean_acc | main    | 6595 | 0.9161 | default |
| KV16           | gpqa_diamond | mean_acc | default | 990  | 0.5081 | default |

**Observation:**  
- On large models (Qwen3-235B-A22B), FP4 maintains accuracy close to FP8/FP16.  
- On smaller models (gpt-oss-120b), FP4 shows more pronounced accuracy drops on difficult datasets.  
- **Trend:** Accuracy degradation is more significant in smaller models.

# Performance Results
Although **speed is not the main goal of this PR** (will be addressed in **#10083 3-2**), we ran throughput tests using **`torch_native`** to provide reference:
- **Reason for `torch_native`:**  
  - Other backends (e.g., `trtllm_mha`, Triton attention) have fused kernels for FP8 only, making FP8 faster there.
  - KV8 lacks a fused kernel on `torch_native`, so both KV4 and KV8 are measured on the same backend.  
  > Note:  KV8 could not run when the attention backend was set to torch_native. We have fixed this problem in PR #12596

- **Test configuration:**  
  - `--num-prompts`: 100–400  
  - `--max-concurrency`: 50–200  
  - Unit: Output token throughput (tok/s)
  
| Num Prompts | Concurrency | KV8 (tok/s) | KV4 (tok/s) | Gain   | TTFT (ms) | TPOT (ms) |
| ----------- | ----------- | ----------- | ----------- | ------ | --------- | --------- |
| 100         | 50          | 62.43       | 60.35       | -3.33% | 5323      | 798       |
| 200         | 100         | 67.34       | 68.02       | +1.0%  | 9378      | 1480      |
| 300         | 150         | 68.81       | 71.63       | +4.1%  | 13500     | 2172      |
| 400         | 200         | 69.75       | 74.19       | +6.36% | 19595     | 2685      |

**Observation:**  
- KV4 shows slightly lower throughput at very small workloads, but catches up and surpasses KV8 at higher concurrency.  
- Speed improvements will be addressed in **#10083 3-2**; this PR focuses on FP4 KV cache feature support.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by JackChuang** [comment]
Hi @Fridge003 @AniZpZ @zhyncs, 
Thank you very much for helping review and merge the PR for MLA KV4 (#10078).
Could you please help review this PR for MLA KV4? Thank you!

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -1314,6 +1314,24 @@ def profile_max_num_token(self, total_gpu_memory: int):
...
+                k = self.model_config.head_dim
+                cell_size = (cell_size // 2) + (
+                    (
+                        n
+                        * k
+                        * num_layers
+                        * 2
+                        * torch._utils._element_size(self.kv_cache_dtype)
+                    )
+                    // scale_block_size
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py:1332]: Why it's not `n * num_layers * 2 * torch._utils._element_size(self.kv_cache_dtype) *  (k // scale_block_size)`, like  mla cache?

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 2]
```diff
@@ -645,24 +645,65 @@ def _create_buffers(self):
...
-                ]
-                self.v_buffer = [
-                    torch.zeros(
-                        (self.size + self.page_size, self.head_num, self.head_dim),
-                        dtype=self.store_dtype,
-                        device=self.device,
-                    )
-                    for _ in range(self.layer_num)
-                ]
+                if is_float4_e2m1fn_x2(self.dtype):
```
Comment (Fridge003) [inline python/sglang/srt/mem_cache/memory_pool.py:648]: Is it possible to overload the MHATokenToKVPool with a `MHATokenToKVPoolFP4`?
There are too many if-else branches here.
I feel the same change needs to be applied to MLA FP4 pool

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/model_executor/model_runner.py (hunk 1 above)
Comment (JackChuang) [reply python/sglang/srt/model_executor/model_runner.py:1332]: @Fridge003 When k is a multiple of scale_block_size (which is usually true for common models), the two formulas are equivalent.
(n * k * num_layers * 2 * torch._utils._element_size(self.kv_cache_dtype) ) // scale_block_size
n * num_layers * 2 * torch._utils._element_size(self.kv_cache_dtype) *  (k // scale_block_size)

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 2 above)
Comment (JackChuang) [reply python/sglang/srt/mem_cache/memory_pool.py:648]: @Fridge003  It’s feasible. How about this — since I have to update the MLA and submit a new PR anyway, let me fix this issue in that PR as well. If you agree with this, I’ll start working on the PR.

Comment (Fridge003) [reply python/sglang/srt/mem_cache/memory_pool.py:648]: Yes, let's open a new PR for it

Comment (JackChuang) [reply python/sglang/srt/mem_cache/memory_pool.py:648]: @Fridge003 Great! I'll create a new PR for the code refactoring for both MLA&MHA token pool fp4.
Meanwhile, could you please help me merge this PR first? This one is for the functionality. I think it’s better to separate the code refactoring from the new features. Thanks~

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
NV tests all passed
https://github.com/sgl-project/sglang/actions/runs/19177871109/job/55448358397?pr=12612

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 2 above)
Comment (JackChuang) [reply python/sglang/srt/mem_cache/memory_pool.py:648]: New PR #13547 has been submitted to refactor the FP4 token pools for both MLA and MHA.
